### DISCUSSION TITLE
"can-of-pure-ability" with MongoAbility

### DISCUSSION THREAD

**Discussion by benjick** [stalniy/casl#818] [Q&A] [answered]
Hello!

What a great project this is, but I'm having some trouble wrapping my head around how to use MongoAbility.

I've read through this doc page, https://casl.js.org/v6/en/api/casl-ability, and some issues/discussions but I'm not any wiser.

First of all, this is my code:
```ts
export const defineUserAbility = (user: User, groupIds: string[]) => {
  const {
    can: allow,
    cannot: forbid,
    build,
  } = new AbilityBuilder(createMongoAbility);
  if (!user.onboarded) {
    return build();
  }
  if (user.role === "TEACHER") {
    allow("create", "Group");
    allow("create", "Task", { groupId: { $in: groupIds } });
    allow("create", "Invite", { groupId: { $in: groupIds } });
  }
  if (user.role === "STUDENT") {
    allow("join", "Group");
  }
  return build();
};
```

From this I then do `const ability = defineUserAbility(user, ["group1", "group2"]);` and I can see with Intellisense that this is a `MongoAbility<AbilityTuple, MongoQuery>` now.

So now in my test I have the following:
```ts
const ability = defineUserAbility(user, ["group1", "group2"]);
expect(ability.can("create", "Group")).toBeTrue(); // works fine
expect(
  ability.can("create", "Task", subject("Task", { groupId: "group1" })),
).toBeTrue();
```

the first one works, but the second one gives me a TypeError: 
```
Argument of type '{ groupId: string; } & ForcedSubject<"Task">' is not assignable to parameter of type 'string | undefined'.
```

and if I ignore the error I get the following output when executing:

> The 3rd, `field` parameter is expected to be a string. See https://stalniy.github.io/casl/en/api/casl-ability#can-of-pure-ability for details

So it feels like it's using the `PureAbility` even though I'm using `createMongoAbility`. I'm obviously missing something.

Any suggestions?

**Comment by stalniy** [comment, marked as answer]
```ts

expect(
  ability.can("create", subject("Task", { groupId: "group1" })),
).toBeTrue();

```

try this ^^^

**Reply by benjick** [reply]
That was it, so obvious now when I see it! 🤦 Thank you!
